In [ ]:
#!pip install pandas  

In [1]:
import pandas as pd

In [2]:
pd.__version__

'2.3.2'

In [3]:
pd.show_versions()


INSTALLED VERSIONS
------------------
commit                : 4665c10899bc413b639194f6fb8665a5c70f7db5
python                : 3.13.5
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_United States.1252

pandas                : 2.3.2
numpy                 : 2.3.3
pytz                  : 2025.2
dateutil              : 2.9.0.post0
pip                   : 25.1
Cython                : None
sphinx                : None
IPython               : 9.1.0
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.13.4
blosc                 : None
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
html5lib            

In [5]:
dir(pd)

['ArrowDtype',
 'BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt8Dtype',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_built_with_meson',
 '_config',
 '_is_numpy_dev',
 '_libs',
 '_pandas_datetime_CAPI',
 '_pandas_parser_CAPI',
 '_testing',
 '_typing',
 '_version_meson',
 'annota

In [6]:
help(pd.qcut)

Help on function qcut in module pandas.core.reshape.tile:

qcut(
    x,
    q,
    labels=None,
    retbins: 'bool' = False,
    precision: 'int' = 3,
    duplicates: 'str' = 'raise'
)
    Quantile-based discretization function.

    Discretize variable into equal-sized buckets based on rank or based
    on sample quantiles. For example 1000 values for 10 quantiles would
    produce a Categorical object indicating quantile membership for each data point.

    Parameters
    ----------
    x : 1d ndarray or Series
    q : int or list-like of float
        Number of quantiles. 10 for deciles, 4 for quartiles, etc. Alternately
        array of quantiles, e.g. [0, .25, .5, .75, 1.] for quartiles.
    labels : array or False, default None
        Used as labels for the resulting bins. Must be of the same length as
        the resulting bins. If False, return only integer indicators of the
        bins. If True, raises an error.
    retbins : bool, optional
        Whether to return the (bin

In [7]:
pd.qcut??

Signature:
pd.qcut(
    x,
    q,
    labels=None,
    retbins: 'bool' = False,
    precision: 'int' = 3,
    duplicates: 'str' = 'raise',
)
Source:   
def qcut(
    x,
    q,
    labels=None,
    retbins: bool = False,
    precision: int = 3,
    duplicates: str = "raise",
):
    """
    Quantile-based discretization function.

    Discretize variable into equal-sized buckets based on rank or based
    on sample quantiles. For example 1000 values for 10 quantiles would
    produce a Categorical object indicating quantile membership for each data point.

    Parameters
    ----------
    x : 1d ndarray or Series
    q : int or list-like of float
        Number of quantiles. 10 for deciles, 4 for quartiles, etc. Alternately
        array of quantiles, e.g. [0, .25, .5, .75, 1.] for quartiles.
    labels : array or False, default None
        Used as labels for the resulting bins. Must be of the same length as
        the resulting bins. If False, return only integer indicators of the
  

# Patient Readmission Analysis

1. Problem statement:
   - Hospital wants to know why patients are readmitted after discharge OR
   - Why are patient readmitted OR
   - Understand why the rate of readmission is increasing
2. Clean the data
3. Transform the data where necessary
4. Analy the data (what drives readmission)
   - Get insights
5. Make Recommendations

In [13]:
#storing the dataset inside a Dataframe
df = pd.read_csv('patient_readmission.csv')

In [14]:
df.shape

(74, 13)

In [15]:
df.head()

patient_id          name    gender  age     city   department     doctor  \
0        3001  Patient_3001      Male   54   Lagos   Orthopedics   Dr. Musa   
1        3002  Patient_3002   female    81    Abuja   Pediatrics   Dr. Musa   
2        3003  Patient_3003      Male   47    Enugu    Neurology  Dr. Smith   
3        3004  Patient_3004      MALE   87   Lagos      Oncology  Dr. Okeke   
4        3005  Patient_3005   female    79    Lagos    Neurology  Dr. Smith   

  admission_date discharge_date bill_amount insurance diagnosis readmitted  
0     2024-06-28     2024-07-11      391724       Yes    Cancer        Yes  
1     2024-06-10     2024-06-18      265621       Yes  Diabetes        Yes  
2     2024-06-16     2024-06-24     487,807       NaN  Fracture        Yes  
3     2024-06-05            NaN      391458       NaN  Diabetes        Yes  
4     2024-06-27     2024-06-30         NaN      yes     Cancer         No

In [12]:
df.tail()

patient_id          name    gender  age     city   department     doctor  \
69        3070  Patient_3070      Male   25    Enugu     Oncology  Dr. Adams   
70        3009  Patient_3009   female    74   Lagos      Oncology  Dr. Okeke   
71        3038  Patient_3038   female    59     Kano   Pediatrics  Dr. Adams   
72        3041  Patient_3041    Female   22    Lagos  Orthopedics  Dr. Okeke   
73        3068  Patient_3068    Female   67   Lagos    Pediatrics  Dr. Smith   

   admission_date discharge_date bill_amount insurance diagnosis readmitted  
69     2024-06-14     2024-06-20      219647        No  Diabetes        Yes  
70            NaN     2024-07-09     388,537      yes   Diabetes        Yes  
71     2024-07-17     2024-07-28     275,039       NaN   cancer          No  
72     2024-07-20     2024-07-26     410,540       NaN  Diabetes         No  
73     2024-07-10     2024-07-24      311076       NaN    Asthma        Yes

In [16]:
df.sample(5)

patient_id          name    gender  age           city   department  \
72        3041  Patient_3041    Female   22          Lagos  Orthopedics   
64        3065  Patient_3065   female    84          Enugu  Orthopedics   
73        3068  Patient_3068    Female   67         Lagos    Pediatrics   
40        3041  Patient_3041    Female   22          Lagos  Orthopedics   
68        3069  Patient_3069      MALE   73  Port Harcourt     Oncology   

       doctor admission_date discharge_date bill_amount insurance diagnosis  \
72  Dr. Okeke     2024-07-20     2024-07-26     410,540       NaN  Diabetes   
64  Dr. Bello            NaN     2024-07-21      427935       NaN  Fracture   
73  Dr. Smith     2024-07-10     2024-07-24      311076       NaN    Asthma   
40  Dr. Okeke     2024-07-20     2024-07-26     410,540       NaN  Diabetes   
68  Dr. Smith     2024-06-28     2024-07-01      428563      yes     Cancer   

   readmitted  
72         No  
64        Yes  
73        Yes  
40         No  
68         No

In [17]:
df

patient_id          name    gender  age     city   department     doctor  \
0         3001  Patient_3001      Male   54   Lagos   Orthopedics   Dr. Musa   
1         3002  Patient_3002   female    81    Abuja   Pediatrics   Dr. Musa   
2         3003  Patient_3003      Male   47    Enugu    Neurology  Dr. Smith   
3         3004  Patient_3004      MALE   87   Lagos      Oncology  Dr. Okeke   
4         3005  Patient_3005   female    79    Lagos    Neurology  Dr. Smith   
..         ...           ...       ...  ...      ...          ...        ...   
69        3070  Patient_3070      Male   25    Enugu     Oncology  Dr. Adams   
70        3009  Patient_3009   female    74   Lagos      Oncology  Dr. Okeke   
71        3038  Patient_3038   female    59     Kano   Pediatrics  Dr. Adams   
72        3041  Patient_3041    Female   22    Lagos  Orthopedics  Dr. Okeke   
73        3068  Patient_3068    Female   67   Lagos    Pediatrics  Dr. Smith   

   admission_date discharge_date bill_amount insurance diagnosis readmitted  
0      2024-06-28     2024-07-11      391724       Yes    Cancer        Yes  
1      2024-06-10     2024-06-18      265621       Yes  Diabetes        Yes  
2      2024-06-16     2024-06-24     487,807       NaN  Fracture        Yes  
3      2024-06-05            NaN      391458       NaN  Diabetes        Yes  
4      2024-06-27     2024-06-30         NaN      yes     Cancer         No  
..            ...            ...         ...       ...       ...        ...  
69     2024-06-14     2024-06-20      219647        No  Diabetes        Yes  
70            NaN     2024-07-09     388,537      yes   Diabetes        Yes  
71     2024-07-17     2024-07-28     275,039       NaN   cancer          No  
72     2024-07-20     2024-07-26     410,540       NaN  Diabetes         No  
73     2024-07-10     2024-07-24      311076       NaN    Asthma        Yes  

[74 rows x 13 columns]

## Let's Explore the Data

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74 entries, 0 to 73
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   patient_id      74 non-null     int64 
 1   name            74 non-null     object
 2   gender          74 non-null     object
 3   age             74 non-null     int64 
 4   city            74 non-null     object
 5   department      74 non-null     object
 6   doctor          74 non-null     object
 7   admission_date  66 non-null     object
 8   discharge_date  69 non-null     object
 9   bill_amount     64 non-null     object
 10  insurance       50 non-null     object
 11  diagnosis       74 non-null     object
 12  readmitted      74 non-null     object
dtypes: int64(2), object(11)
memory usage: 7.6+ KB


In [19]:
# check for duplicates
df.duplicated().any()

np.True_

In [20]:
df.duplicated()

0     False
1     False
2     False
3     False
4     False
      ...  
69    False
70     True
71     True
72     True
73     True
Length: 74, dtype: bool

In [21]:
df.shape

(74, 13)

In [22]:
df.duplicated().sum()

np.int64(4)

In [25]:
#displays the section of the df that has duplicates
df[df.duplicated()]

patient_id          name    gender  age     city   department     doctor  \
70        3009  Patient_3009   female    74   Lagos      Oncology  Dr. Okeke   
71        3038  Patient_3038   female    59     Kano   Pediatrics  Dr. Adams   
72        3041  Patient_3041    Female   22    Lagos  Orthopedics  Dr. Okeke   
73        3068  Patient_3068    Female   67   Lagos    Pediatrics  Dr. Smith   

   admission_date discharge_date bill_amount insurance diagnosis readmitted  
70            NaN     2024-07-09     388,537      yes   Diabetes        Yes  
71     2024-07-17     2024-07-28     275,039       NaN   cancer          No  
72     2024-07-20     2024-07-26     410,540       NaN  Diabetes         No  
73     2024-07-10     2024-07-24      311076       NaN    Asthma        Yes

In [29]:
df.drop_duplicates(inplace=True)
# inplace = True :  Modifies the DataFrame directly instead of returning a new one

In [28]:
df

patient_id          name    gender  age           city    department  \
0         3001  Patient_3001      Male   54         Lagos    Orthopedics   
1         3002  Patient_3002   female    81          Abuja    Pediatrics   
2         3003  Patient_3003      Male   47          Enugu     Neurology   
3         3004  Patient_3004      MALE   87         Lagos       Oncology   
4         3005  Patient_3005   female    79          Lagos     Neurology   
..         ...           ...       ...  ...            ...           ...   
65        3066  Patient_3066      Male   33          Lagos     Neurology   
66        3067  Patient_3067   female    39           Kano   cardiology    
67        3068  Patient_3068    Female   67         Lagos     Pediatrics   
68        3069  Patient_3069      MALE   73  Port Harcourt      Oncology   
69        3070  Patient_3070      Male   25          Enugu      Oncology   

       doctor admission_date discharge_date bill_amount insurance diagnosis  \
0    Dr. Musa     2024-06-28     2024-07-11      391724       Yes    Cancer   
1    Dr. Musa     2024-06-10     2024-06-18      265621       Yes  Diabetes   
2   Dr. Smith     2024-06-16     2024-06-24     487,807       NaN  Fracture   
3   Dr. Okeke     2024-06-05            NaN      391458       NaN  Diabetes   
4   Dr. Smith     2024-06-27     2024-06-30         NaN      yes     Cancer   
..        ...            ...            ...         ...       ...       ...   
65  Dr. Bello     2024-07-25     2024-07-28       66168        No   cancer    
66  Dr. Bello     2024-07-24     2024-08-10         NaN        No   cancer    
67  Dr. Smith     2024-07-10     2024-07-24      311076       NaN    Asthma   
68  Dr. Smith     2024-06-28     2024-07-01      428563      yes     Cancer   
69  Dr. Adams     2024-06-14     2024-06-20      219647        No  Diabetes   

   readmitted  
0         Yes  
1         Yes  
2         Yes  
3         Yes  
4          No  
..        ...  
65        Yes  
66        Yes  
67        Yes  
68         No  
69        Yes  

[70 rows x 13 columns]

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70 entries, 0 to 69
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   patient_id      70 non-null     int64 
 1   name            70 non-null     object
 2   gender          70 non-null     object
 3   age             70 non-null     int64 
 4   city            70 non-null     object
 5   department      70 non-null     object
 6   doctor          70 non-null     object
 7   admission_date  63 non-null     object
 8   discharge_date  65 non-null     object
 9   bill_amount     60 non-null     object
 10  insurance       49 non-null     object
 11  diagnosis       70 non-null     object
 12  readmitted      70 non-null     object
dtypes: int64(2), object(11)
memory usage: 7.7+ KB


In [31]:
df

patient_id          name    gender  age           city    department  \
0         3001  Patient_3001      Male   54         Lagos    Orthopedics   
1         3002  Patient_3002   female    81          Abuja    Pediatrics   
2         3003  Patient_3003      Male   47          Enugu     Neurology   
3         3004  Patient_3004      MALE   87         Lagos       Oncology   
4         3005  Patient_3005   female    79          Lagos     Neurology   
..         ...           ...       ...  ...            ...           ...   
65        3066  Patient_3066      Male   33          Lagos     Neurology   
66        3067  Patient_3067   female    39           Kano   cardiology    
67        3068  Patient_3068    Female   67         Lagos     Pediatrics   
68        3069  Patient_3069      MALE   73  Port Harcourt      Oncology   
69        3070  Patient_3070      Male   25          Enugu      Oncology   

       doctor admission_date discharge_date bill_amount insurance diagnosis  \
0    Dr. Musa     2024-06-28     2024-07-11      391724       Yes    Cancer   
1    Dr. Musa     2024-06-10     2024-06-18      265621       Yes  Diabetes   
2   Dr. Smith     2024-06-16     2024-06-24     487,807       NaN  Fracture   
3   Dr. Okeke     2024-06-05            NaN      391458       NaN  Diabetes   
4   Dr. Smith     2024-06-27     2024-06-30         NaN      yes     Cancer   
..        ...            ...            ...         ...       ...       ...   
65  Dr. Bello     2024-07-25     2024-07-28       66168        No   cancer    
66  Dr. Bello     2024-07-24     2024-08-10         NaN        No   cancer    
67  Dr. Smith     2024-07-10     2024-07-24      311076       NaN    Asthma   
68  Dr. Smith     2024-06-28     2024-07-01      428563      yes     Cancer   
69  Dr. Adams     2024-06-14     2024-06-20      219647        No  Diabetes   

   readmitted  
0         Yes  
1         Yes  
2         Yes  
3         Yes  
4          No  
..        ...  
65        Yes  
66        Yes  
67        Yes  
68         No  
69        Yes  

[70 rows x 13 columns]

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 70 entries, 0 to 69
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   patient_id      70 non-null     int64 
 1   name            70 non-null     object
 2   gender          70 non-null     object
 3   age             70 non-null     int64 
 4   city            70 non-null     object
 5   department      70 non-null     object
 6   doctor          70 non-null     object
 7   admission_date  63 non-null     object
 8   discharge_date  65 non-null     object
 9   bill_amount     60 non-null     object
 10  insurance       49 non-null     object
 11  diagnosis       70 non-null     object
 12  readmitted      70 non-null     object
dtypes: int64(2), object(11)
memory usage: 7.7+ KB


In [35]:
df.isnull().sum()

patient_id         0
name               0
gender             0
age                0
city               0
department         0
doctor             0
admission_date     7
discharge_date     5
bill_amount       10
insurance         21
diagnosis          0
readmitted         0
dtype: int64

In [38]:
df.shape

(70, 13)

In [44]:
#calculates the number of rows remaining in the DataFrame after all rows with missing values have been excluded.
#in short form - return the number of rows without missing values
len(df.dropna())

37

In [37]:
df.head()

patient_id          name    gender  age     city   department     doctor  \
0        3001  Patient_3001      Male   54   Lagos   Orthopedics   Dr. Musa   
1        3002  Patient_3002   female    81    Abuja   Pediatrics   Dr. Musa   
2        3003  Patient_3003      Male   47    Enugu    Neurology  Dr. Smith   
3        3004  Patient_3004      MALE   87   Lagos      Oncology  Dr. Okeke   
4        3005  Patient_3005   female    79    Lagos    Neurology  Dr. Smith   

  admission_date discharge_date bill_amount insurance diagnosis readmitted  
0     2024-06-28     2024-07-11      391724       Yes    Cancer        Yes  
1     2024-06-10     2024-06-18      265621       Yes  Diabetes        Yes  
2     2024-06-16     2024-06-24     487,807       NaN  Fracture        Yes  
3     2024-06-05            NaN      391458       NaN  Diabetes        Yes  
4     2024-06-27     2024-06-30         NaN      yes     Cancer         No

In [39]:
len(df.dropna(subset=['admission_date', 'discharge_date']))

58

In [41]:
df.dropna(subset=['admission_date', 'discharge_date'], inplace=True)

In [42]:
df.shape

(58, 13)

In [48]:
#output shows number of rows left in the dataframe - look at the index eg row 3 has been removed
df

patient_id          name    gender  age           city    department  \
0         3001  Patient_3001      Male   54         Lagos    Orthopedics   
1         3002  Patient_3002   female    81          Abuja    Pediatrics   
2         3003  Patient_3003      Male   47          Enugu     Neurology   
4         3005  Patient_3005   female    79          Lagos     Neurology   
5         3006  Patient_3006      Male   40           Kano      Oncology   
6         3007  Patient_3007      Male   37          Enugu   cardiology    
7         3008  Patient_3008      MALE   86         Lagos     Cardiology   
9         3010  Patient_3010   female    66         Lagos       Oncology   
10        3011  Patient_3011      Male   77         Lagos     Pediatrics   
12        3013  Patient_3013    Female   19          Enugu   cardiology    
13        3014  Patient_3014   female    35           Kano      Oncology   
14        3015  Patient_3015   female    76          Enugu    Cardiology   
15        3016  Patient_3016    Female   42          Lagos   cardiology    
16        3017  Patient_3017    Female   20          Abuja    Cardiology   
17        3018  Patient_3018      Male   50  Port Harcourt   Orthopedics   
19        3020  Patient_3020      Male   52  Port Harcourt   Orthopedics   
20        3021  Patient_3021      MALE   73  Port Harcourt   cardiology    
22        3023  Patient_3023      Male   18          Abuja   Orthopedics   
23        3024  Patient_3024   female    31           Kano   Orthopedics   
24        3025  Patient_3025      MALE   26  Port Harcourt     Neurology   
25        3026  Patient_3026      Male   83          Lagos    Pediatrics   
26        3027  Patient_3027      Male   63  Port Harcourt   Orthopedics   
28        3029  Patient_3029    Female   28           Kano     Neurology   
29        3030  Patient_3030    Female   73          Abuja    Pediatrics   
30        3031  Patient_3031      Male   19  Port Harcourt    Cardiology   
31        3032  Patient_3032   female    86           Kano    Pediatrics   
32        3033  Patient_3033      MALE   66          Lagos    Pediatrics   
33        3034  Patient_3034      MALE   27           Kano      Oncology   
34        3035  Patient_3035   female    87  Port Harcourt   Orthopedics   
35        3036  Patient_3036      MALE   69          Abuja     Neurology   
36        3037  Patient_3037      Male   21          Abuja      Oncology   
37        3038  Patient_3038   female    59           Kano    Pediatrics   
38        3039  Patient_3039   female    51           Kano   Orthopedics   
40        3041  Patient_3041    Female   22          Lagos   Orthopedics   
41        3042  Patient_3042   female    60  Port Harcourt    Cardiology   
43        3044  Patient_3044      MALE   62          Abuja      Oncology   
44        3045  Patient_3045      MALE   21  Port Harcourt      Oncology   
45        3046  Patient_3046      MALE   38          Abuja   Orthopedics   
47        3048  Patient_3048   female    26          Abuja    Pediatrics   
48        3049  Patient_3049      Male   36           Kano    Cardiology   
49        3050  Patient_3050    Female   65  Port Harcourt      Oncology   
51        3052  Patient_3052      MALE   28          Lagos   Orthopedics   
52        3053  Patient_3053   female    25  Port Harcourt    Cardiology   
53        3054  Patient_3054    Female   84           Kano    Pediatrics   
54        3055  Patient_3055   female    67          Lagos      Oncology   
55        3056  Patient_3056   female    86           Kano   cardiology    
56        3057  Patient_3057      MALE   57  Port Harcourt   Orthopedics   
57        3058  Patient_3058      Male   36  Port Harcourt     Neurology   
59        3060  Patient_3060      MALE   67  Port Harcourt    Pediatrics   
60        3061  Patient_3061   female    65          Abuja   Orthopedics   
61        3062  Patient_3062      Male   34          Enugu    Pediatrics   
62        3063  Patient_3063    Female   86         

In [50]:
# resets the index of the df to the default integer index eg (0,1,2..),while discarding old index (above output)
df.reset_index(drop=True)

patient_id          name    gender  age           city    department  \
0         3001  Patient_3001      Male   54         Lagos    Orthopedics   
1         3002  Patient_3002   female    81          Abuja    Pediatrics   
2         3003  Patient_3003      Male   47          Enugu     Neurology   
3         3005  Patient_3005   female    79          Lagos     Neurology   
4         3006  Patient_3006      Male   40           Kano      Oncology   
5         3007  Patient_3007      Male   37          Enugu   cardiology    
6         3008  Patient_3008      MALE   86         Lagos     Cardiology   
7         3010  Patient_3010   female    66         Lagos       Oncology   
8         3011  Patient_3011      Male   77         Lagos     Pediatrics   
9         3013  Patient_3013    Female   19          Enugu   cardiology    
10        3014  Patient_3014   female    35           Kano      Oncology   
11        3015  Patient_3015   female    76          Enugu    Cardiology   
12        3016  Patient_3016    Female   42          Lagos   cardiology    
13        3017  Patient_3017    Female   20          Abuja    Cardiology   
14        3018  Patient_3018      Male   50  Port Harcourt   Orthopedics   
15        3020  Patient_3020      Male   52  Port Harcourt   Orthopedics   
16        3021  Patient_3021      MALE   73  Port Harcourt   cardiology    
17        3023  Patient_3023      Male   18          Abuja   Orthopedics   
18        3024  Patient_3024   female    31           Kano   Orthopedics   
19        3025  Patient_3025      MALE   26  Port Harcourt     Neurology   
20        3026  Patient_3026      Male   83          Lagos    Pediatrics   
21        3027  Patient_3027      Male   63  Port Harcourt   Orthopedics   
22        3029  Patient_3029    Female   28           Kano     Neurology   
23        3030  Patient_3030    Female   73          Abuja    Pediatrics   
24        3031  Patient_3031      Male   19  Port Harcourt    Cardiology   
25        3032  Patient_3032   female    86           Kano    Pediatrics   
26        3033  Patient_3033      MALE   66          Lagos    Pediatrics   
27        3034  Patient_3034      MALE   27           Kano      Oncology   
28        3035  Patient_3035   female    87  Port Harcourt   Orthopedics   
29        3036  Patient_3036      MALE   69          Abuja     Neurology   
30        3037  Patient_3037      Male   21          Abuja      Oncology   
31        3038  Patient_3038   female    59           Kano    Pediatrics   
32        3039  Patient_3039   female    51           Kano   Orthopedics   
33        3041  Patient_3041    Female   22          Lagos   Orthopedics   
34        3042  Patient_3042   female    60  Port Harcourt    Cardiology   
35        3044  Patient_3044      MALE   62          Abuja      Oncology   
36        3045  Patient_3045      MALE   21  Port Harcourt      Oncology   
37        3046  Patient_3046      MALE   38          Abuja   Orthopedics   
38        3048  Patient_3048   female    26          Abuja    Pediatrics   
39        3049  Patient_3049      Male   36           Kano    Cardiology   
40        3050  Patient_3050    Female   65  Port Harcourt      Oncology   
41        3052  Patient_3052      MALE   28          Lagos   Orthopedics   
42        3053  Patient_3053   female    25  Port Harcourt    Cardiology   
43        3054  Patient_3054    Female   84           Kano    Pediatrics   
44        3055  Patient_3055   female    67          Lagos      Oncology   
45        3056  Patient_3056   female    86           Kano   cardiology    
46        3057  Patient_3057      MALE   57  Port Harcourt   Orthopedics   
47        3058  Patient_3058      Male   36  Port Harcourt     Neurology   
48        3060  Patient_3060      MALE   67  Port Harcourt    Pediatrics   
49        3061  Patient_3061   female    65          Abuja   Orthopedics   
50        3062  Patient_3062      Male   34          Enugu    Pediatrics   
51        3063  Patient_3063    Female   86         

In [53]:
df.shape

(58, 13)

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 69
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   patient_id      58 non-null     int64 
 1   name            58 non-null     object
 2   gender          58 non-null     object
 3   age             58 non-null     int64 
 4   city            58 non-null     object
 5   department      58 non-null     object
 6   doctor          58 non-null     object
 7   admission_date  58 non-null     object
 8   discharge_date  58 non-null     object
 9   bill_amount     51 non-null     object
 10  insurance       42 non-null     object
 11  diagnosis       58 non-null     object
 12  readmitted      58 non-null     object
dtypes: int64(2), object(11)
memory usage: 6.3+ KB


In [52]:
df.columns

Index(['patient_id', 'name', 'gender', 'age', 'city', 'department', 'doctor',
       'admission_date', 'discharge_date', 'bill_amount', 'insurance',
       'diagnosis', 'readmitted'],
      dtype='object')

In [54]:
# convert data column from object to date data type
df['admission_date'] = pd.to_datetime(df['admission_date'])
df['discharge_date'] = pd.to_datetime(df['discharge_date'])

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58 entries, 0 to 69
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   patient_id      58 non-null     int64         
 1   name            58 non-null     object        
 2   gender          58 non-null     object        
 3   age             58 non-null     int64         
 4   city            58 non-null     object        
 5   department      58 non-null     object        
 6   doctor          58 non-null     object        
 7   admission_date  58 non-null     datetime64[ns]
 8   discharge_date  58 non-null     datetime64[ns]
 9   bill_amount     51 non-null     object        
 10  insurance       42 non-null     object        
 11  diagnosis       58 non-null     object        
 12  readmitted      58 non-null     object        
dtypes: datetime64[ns](2), int64(2), object(9)
memory usage: 6.3+ KB


In [56]:
df["admission_days"]  = (df['discharge_date'] - df['admission_date']).dt.days

In [57]:
df.head()

patient_id          name    gender  age     city   department     doctor  \
0        3001  Patient_3001      Male   54   Lagos   Orthopedics   Dr. Musa   
1        3002  Patient_3002   female    81    Abuja   Pediatrics   Dr. Musa   
2        3003  Patient_3003      Male   47    Enugu    Neurology  Dr. Smith   
4        3005  Patient_3005   female    79    Lagos    Neurology  Dr. Smith   
5        3006  Patient_3006      Male   40     Kano     Oncology  Dr. Smith   

  admission_date discharge_date bill_amount insurance diagnosis readmitted  \
0     2024-06-28     2024-07-11      391724       Yes    Cancer        Yes   
1     2024-06-10     2024-06-18      265621       Yes  Diabetes        Yes   
2     2024-06-16     2024-06-24     487,807       NaN  Fracture        Yes   
4     2024-06-27     2024-06-30         NaN      yes     Cancer         No   
5     2024-07-30     2024-08-14       46244       Yes    Cancer        Yes   

   admission_days  
0              13  
1               8  
2               8  
4               3  
5              15

## Simple analysis

In [60]:
(df["readmitted"].value_counts(normalize=True) * 100).round(2)

readmitted
No     53.45
Yes    46.55
Name: proportion, dtype: float64

#### Average stay in the hospital

In [61]:
df.groupby("readmitted")["admission_days"].mean()

readmitted
No     8.161290
Yes    9.777778
Name: admission_days, dtype: float64

#### By Diagnosis

In [64]:
#percentage rate of patients readmitted based on their diagnosis
(df.groupby("diagnosis")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted      No   Yes
diagnosis               
Hypertension  11.0  89.0
Diabetes      56.0  44.0
 cancer       57.0  43.0
Cancer        58.0  42.0
Asthma        71.0  29.0
Fracture      71.0  29.0

In [65]:
(df.groupby("gender")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted    No   Yes
gender                
Male        39.0  61.0
 female     59.0  41.0
Female      60.0  40.0
MALE        62.0  38.0

In [67]:
(df.groupby("age")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round().head()

readmitted  No    Yes
age                  
18         NaN  100.0
37         NaN  100.0
81         NaN  100.0
57         NaN  100.0
54         NaN  100.0

In [68]:
(df.groupby("department")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted      No   Yes
department              
 cardiology   17.0  83.0
Neurology     43.0  57.0
Oncology      54.0  46.0
Orthopedics   54.0  46.0
Pediatrics    67.0  33.0
Cardiology    71.0  29.0

##### >>Based on outputs above more data cleaning is required for more accurate insights 

In [ ]:
#### More Da

In [69]:
# Clean the gender column
df["gender"] = df["gender"].str.strip().str.lower()

In [70]:
(df.groupby("gender")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted    No   Yes
gender                
male        48.0  52.0
female      59.0  41.0

In [71]:
# Clean the department column
df["department"] = df["department"].str.strip().str.title()

In [72]:
(df.groupby("department")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted     No   Yes
department             
Neurology    43.0  57.0
Cardiology   46.0  54.0
Oncology     54.0  46.0
Orthopedics  54.0  46.0
Pediatrics   67.0  33.0

In [73]:
# Clean the Diagnosis column
df["diagnosis"] = df["diagnosis"].str.strip().str.title()

In [74]:
(df.groupby("diagnosis")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted      No   Yes
diagnosis               
Hypertension  11.0  89.0
Diabetes      56.0  44.0
Cancer        58.0  42.0
Asthma        71.0  29.0
Fracture      71.0  29.0

In [80]:
# Create a new column for age group 

# Define a function for age grouping
def age_group(age):
    if age <= 30:
        return "Youth (≤30)"
    elif age <= 45:
        return "Adult (31-45)"
    elif age <= 60:
        return "Middle Age (46-60)"
    else:
        return "Senior (61+)"



In [81]:
# Apply the function to create a new column
df["age_group"] = df["age"].apply(age_group)

In [82]:
(df.groupby("age_group")["readmitted"].value_counts(normalize=True).unstack().sort_values("Yes", ascending=False) * 100).round()

readmitted            No   Yes
age_group                     
Adult (31-45)       27.0  73.0
Youth (≤30)         36.0  64.0
Middle Age (46-60)  50.0  50.0
Senior (61+)        76.0  24.0

## 📊 Insights from the Analysis

### 1. Overall Readmission Rate
- **46.6% of patients were readmitted**.
- That’s nearly half — a significant challenge for the hospital.

### 2. Length of Stay
- Readmitted patients stayed **longer on average (≈9.8 days)** compared to those not readmitted (≈8.2 days).
- Longer initial stays may indicate more severe conditions or complications.

### 3. By Department
- **Neurology (57%)** and **Cardiology (54%)** have the highest readmission rates.
- **Oncology (46%)** and **Orthopedics (46%)** are more balanced.
- **Pediatrics (33%)** has the lowest readmission rate.

### 4. By Diagnosis
- **Hypertension**: extremely high readmission at **89%** — an urgent red flag.
- **Diabetes (44%)** and **Cancer (42%)** patients also face moderate readmission risk.
- **Asthma (29%)** and **Fracture (29%)** have the lowest readmission rates.

### 5. By Age Group
- **Adults (31–45)**: highest at **73% readmission**.
- **Youth (≤30)**: **64% readmitted**.
- **Middle Age (46–60)**: balanced at **50%**.
- **Seniors (61+)**: lowest at **24%**, which is unusual and may reflect sample size or mortality bias.

---

## 💡 Recommendations

1. **Focus on chronic conditions**: Provide better follow-up and treatment plans for Hypertension and Diabetes patients.
2. **Departmental reviews**: Cardiology and Neurology should strengthen discharge planning and patient education.
3. **Age-specific interventions**: Target young and middle-aged adults with lifestyle and adherence programs.
4. **Support for long-stay patients**: Provide additional discharge checks and care plans to reduce bounce-back readmissions.
